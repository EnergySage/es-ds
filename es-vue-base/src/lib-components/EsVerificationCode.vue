<template>
    <b-input-group class="code-holder">
        <b-form-input
            v-for="(n, index) in code"
            :id="`codeInput_${index}`"
            :ref="`codeInput_${index}`"
            :key="index"
            novalidate
            :value="code[index]"
            :name="'codeInput_' + index"
            :type="type"
            :pattern="pattern"
            class="code-input text-center"
            maxlength="1"
            autocomplete="off"
            aria-autocomplete="none"
            :aria-label="`digit ${index} of ${code.length}`"
            v-bind="$attrs"
            @input.native="handleInput"
            @keypress="handleKeys"
            @keyup.enter="emitCodeUpdate"
            @keydown.delete="handleDelete"
            @keydown.left="handleLeft"
            @keydown.right="handleRight"
            @paste="handlePaste" />
    </b-input-group>
</template>

<script lang="js">
import { BInputGroup, BFormInput } from 'bootstrap-vue';

export default {
    name: 'EsVerificationCode',
    components: {
        BInputGroup,
        BFormInput,
    },
    props: {
        /**
         * Character Count
         */
        charCount: {
            type: Number,
            default: 5,
        },
        /**
         * Allowed characters to type
         * Defaults to 0-9
         */
        allowedChars: {
            type: Array,
            default: () => Array(10).fill(0).map((e, i) => `${i}`),
        },
        /**
         * Input Type
         */
        type: {
            type: String,
            // tel causes mobile to show numeric keyboard
            default: 'tel',
        },
        /**
         * Input Pattern
         */
        pattern: {
            type: String,
            default: '[0-9]*',
        },
        /**
         * Code Value From Parent
         */
        value: {
            type: Array,
            required: true,
        },
    },
    data() {
        return {
            dataFromPaste: '',
            emptyArray: Array(this.charCount).fill(''),
            code: Array(this.charCount).fill(''),
        };
    },
    watch: {
        // Watch for parent component changes to value
        value: {
            deep: true,
            handler(newVal, oldVal) {
                // On value update from parent make sure it is not the result of emit('input')
                if (newVal.toString() !== oldVal.toString()) {
                    this.valuePropChange(newVal);
                }
            },
        },
    },
    created() {
        if (this.value) {
            // On first create if a value is set update this.code
            this.valuePropChange(this.value);
        }
    },
    methods: {
        // Returns the correct this.code index based on DOM element
        getElementIndex(element) {
            return parseInt(element?.id?.split('_')?.[1] || -1, 10);
        },
        // Triggers on keypress outside of [left, right, delete, enter]
        handleKeys(event) {
            // Do not allow for keys not in allowed except ctrl + cmd
            if (!event.ctrlKey && !event.metaKey && !this.allowedChars.includes(event.key)) {
                event.preventDefault();
            }
        },
        // On left arrow move to previous input
        handleLeft(event) {
            const currentInput = event.target;
            const previousInput = currentInput.previousElementSibling;
            const cursorPosition = currentInput.selectionStart;

            if (previousInput && cursorPosition === 0) {
                event.preventDefault();
                previousInput?.focus();
                previousInput.setSelectionRange(-1, -1);
            }
        },
        // On right arrow move to next input
        handleRight(event) {
            const currentInput = event.target;
            const currentValue = currentInput.value;
            const nextInput = currentInput.nextElementSibling;
            const cursorPosition = currentInput.selectionStart;

            if (nextInput && (cursorPosition === 1 || currentValue.length === 0)) {
                event.preventDefault();
                nextInput?.focus();
            }
        },
        // On delete in an empty input remove previous input value if possible
        handleDelete(event) {
            const { value } = event.target;
            const currentInput = event.target;
            const previousInput = currentInput.previousElementSibling;
            const currentIndex = this.getElementIndex(currentInput);
            const prevIndex = this.getElementIndex(previousInput);

            // If delete is pressed on an empty input
            if (!value) {
                // Delete the previous inputs value if possible
                if (prevIndex >= 0) {
                    // We use $set because replacing by index can remove reactivity
                    this.$set(this.code, prevIndex, '');
                    previousInput?.focus();
                }
            // If not clear the current input
            } else {
                this.$set(this.code, currentIndex, '');
            }

            this.emitCodeUpdate();
        },
        handleInput(event) {
            const { inputType } = event;
            const currentActiveElement = event.target;
            // Getting the current index handles case where user pastes into an input other than input[0]
            const curIndex = this.getElementIndex(currentActiveElement);

            // Triggers when a character is added only
            if (inputType === 'insertText') {
                // We use $set because replacing by index can remove reactivity
                this.$set(this.code, curIndex, event.target.value);
                // Move to next input if possible
                currentActiveElement.nextElementSibling?.focus();
            }

            // Always emit input events; even if nothing changes
            this.emitCodeUpdate();
        },
        handlePaste(event) {
            // Get the paste from clipboard; truncate it if its too long
            this.dataFromPaste = event.clipboardData?.getData('text')
                ?.trim()?.split('')?.splice(0, this.charCount);

            // If no data is found or clipboard api is not supported stop
            if (!this.dataFromPaste) {
                return;
            }

            // Empty existing code and fill it with pasted data
            this.code = [...this.emptyArray].map((cur, index) => this.dataFromPaste[index] || '');

            // Move focus to the next empty spot or last input
            let focusIndex = this.dataFromPaste.length;
            if (focusIndex > this.code.length - 1) {
                focusIndex = this.code.length - 1;
            }
            this.$refs[`codeInput_${focusIndex}`]?.[0]?.focus();

            event.preventDefault();
            this.emitCodeUpdate();
        },
        // If parent value changes normalize it and trigger updates
        valuePropChange(value) {
            this.code = [...this.emptyArray].map((cur, index) => value[index] || '');

            this.emitCodeUpdate(false);
        },
        // Emit this.code changes to keep this.value in sync
        emitCodeUpdate(emitInput = true) {
            const codeIsValid = this.code.every((num) => {
                if (!this.allowedChars.includes(num)) {
                    return false;
                }
                return true;
            }) && this.code.length === this.charCount;

            if (emitInput) {
                this.$emit('input', this.code);
            }
            this.$emit('valid-code', codeIsValid);
        },
    },
};
</script>

<style lang="scss">
@import '~@energysage/es-bs-base/scss/includes';

.code-holder {
    margin: 0 auto;
    max-width: 500px;

    .code-input {
        appearance: textfield;
        font-size: 3rem;
        height: 6rem;

        &::-webkit-outer-spin-button,
        &::-webkit-inner-spin-button {
            appearance: none;
        }
    }
}
</style>
