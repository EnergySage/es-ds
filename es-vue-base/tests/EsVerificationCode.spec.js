import { mount } from '@vue/test-utils';
import EsVerificationCode from '@/src/lib-components/EsVerificationCode.vue';
import jestVue from '@/tests/jest.vue.config';

describe('EsVerificationCode', () => {
    beforeAll(() => {
        // JSDom does not implement this and an error was being
        // thrown from jest-axe because of it.
        window.getComputedStyle = () => {};
    });
    // Basic test; should exist for most components
    test('<EsVerificationCode />', async () => {
        const wrapper = mount(EsVerificationCode, {
            ...jestVue,
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();
    });

    test('<EsVerificationCode has the correct number of inputs />', async () => {
        const charCount = 5;
        const wrapper = mount(EsVerificationCode, {
            ...jestVue,
            propsData: {
                charCount,
            },
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.findAll('.code-input')).toHaveLength(charCount);
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();
    });

    test('<EsVerificationCode emits input />', async () => {
        const wrapper = mount(EsVerificationCode, {
            ...jestVue,
        });
        const a11y = await axe(wrapper.element);
        const fisrtInput = wrapper.find('.code-input');

        fisrtInput.element.value = 1;
        fisrtInput.trigger('input');

        expect(wrapper.emitted().input).toEqual([[['1', '', '', '', '']]]);
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();
    });

    test('<EsVerificationCode emits valid-code false when invalid />', async () => {
        const wrapper = mount(EsVerificationCode, {
            ...jestVue,
        });
        const a11y = await axe(wrapper.element);
        const codeInputs = wrapper.findAll('.code-input');
        const chars = ['a', '/', '+', 'B', '?'];

        codeInputs.wrappers.forEach((input, index) => {
            // eslint-disable-next-line no-param-reassign
            input.element.value = chars[index];
            input.trigger('input');
        });

        expect(wrapper.emitted()['valid-code']).toEqual([[false], [false], [false], [false], [false]]);
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();
    });

    test('<EsVerificationCode emits valid-code true when valid />', async () => {
        const wrapper = mount(EsVerificationCode, {
            ...jestVue,
        });
        const a11y = await axe(wrapper.element);
        const codeInputs = wrapper.findAll('.code-input');

        codeInputs.wrappers.forEach((input) => {
            // eslint-disable-next-line no-param-reassign
            input.element.value = 1;
            input.trigger('input');
        });

        expect(wrapper.emitted()['valid-code']).toEqual([[false], [false], [false], [false], [true]]);
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();
    });

    // TODO: Simulate Paste
});
