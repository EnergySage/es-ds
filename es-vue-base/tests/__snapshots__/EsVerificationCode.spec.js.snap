// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsVerificationCode <EsVerificationCode /> 1`] = `
"<div role="group" class="input-group code-holder justify-content-center">
  <!----><input id="codeInput_0" name="codeInput_0" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 0 of 5"><input id="codeInput_1" name="codeInput_1" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 1 of 5"><input id="codeInput_2" name="codeInput_2" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 2 of 5"><input id="codeInput_3" name="codeInput_3" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 3 of 5"><input id="codeInput_4" name="codeInput_4" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 4 of 5">
  <!---->
</div>"
`;

exports[`EsVerificationCode <EsVerificationCode emits input /> 1`] = `
"<div role="group" class="input-group code-holder justify-content-center">
  <!----><input id="codeInput_0" name="codeInput_0" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 0 of 5"><input id="codeInput_1" name="codeInput_1" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 1 of 5"><input id="codeInput_2" name="codeInput_2" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 2 of 5"><input id="codeInput_3" name="codeInput_3" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 3 of 5"><input id="codeInput_4" name="codeInput_4" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 4 of 5">
  <!---->
</div>"
`;

exports[`EsVerificationCode <EsVerificationCode emits valid-code false when invalid /> 1`] = `
"<div role="group" class="input-group code-holder justify-content-center">
  <!----><input id="codeInput_0" name="codeInput_0" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 0 of 5"><input id="codeInput_1" name="codeInput_1" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 1 of 5"><input id="codeInput_2" name="codeInput_2" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 2 of 5"><input id="codeInput_3" name="codeInput_3" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 3 of 5"><input id="codeInput_4" name="codeInput_4" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 4 of 5">
  <!---->
</div>"
`;

exports[`EsVerificationCode <EsVerificationCode emits valid-code true when valid /> 1`] = `
"<div role="group" class="input-group code-holder justify-content-center">
  <!----><input id="codeInput_0" name="codeInput_0" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 0 of 5"><input id="codeInput_1" name="codeInput_1" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 1 of 5"><input id="codeInput_2" name="codeInput_2" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 2 of 5"><input id="codeInput_3" name="codeInput_3" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 3 of 5"><input id="codeInput_4" name="codeInput_4" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 4 of 5">
  <!---->
</div>"
`;

exports[`EsVerificationCode <EsVerificationCode has the correct number of inputs /> 1`] = `
"<div role="group" class="input-group code-holder justify-content-center">
  <!----><input id="codeInput_0" name="codeInput_0" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 0 of 5"><input id="codeInput_1" name="codeInput_1" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 1 of 5"><input id="codeInput_2" name="codeInput_2" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 2 of 5"><input id="codeInput_3" name="codeInput_3" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 3 of 5"><input id="codeInput_4" name="codeInput_4" type="tel" autocomplete="off" class="code-input text-center font-size-300 font-size-lg-500 mr-50 mr-lg-100 font-weight-bold px-100 px-lg-150 py-150 form-control" novalidate="novalidate" pattern="[0-9]*" maxlength="1" aria-autocomplete="none" aria-label="digit 4 of 5">
  <!---->
</div>"
`;
